### ISSUE TITLE
Private Keys Vulnerable to New Android Attack

### ISSUE THREAD

**Issue by HashEngineering** [bitcoinj/bitcoinj#1211]
There is a new article out concerning this matter.

https://news.bitcoin.com/private-keys-vulnerable-to-new-android-ios-attack/

The above article refers to these two papers:
1. https://www.cs.tau.ac.il/~tromer/mobilesc/
2. https://eprint.iacr.org/2016/231

Paper 1 says:
"BouncyCastle: Android's version of BouncyCastle is vulnerable, as discovered by a concurrent and independent work, Side-Channel Analysis of Weierstrass and Koblitz Curve ECDSA on Android Smartphones."

Is this a matter to be concerned about?  Is there a plan for code changes?

**Comment by schildbach** [comment]
I'm quite sure any countermeasures will need to be implemented in [BouncyCastle](http://bouncycastle.org/)/[SpongyCastle](https://github.com/rtyley/spongycastle).

However, the paper says

> The attack requires measuring a few thousand ECDSA signatures

…presumably by the same key. This means massive address re-use, which bitcoinj already disregards by using HD wallets etc. So in a way, we implemented the countermeasure already years ago.

**Comment by greenaddress** [comment]
As of early February secp256k1 comes with tested [JNI bindings](https://github.com/bitcoin/secp256k1/commit/e4570184ffba0794dee0dd6f5d76d4edeb50fa0d)

Both openssl an BC are likely to patch the code against this vulnerability; even so, the objective-C library Core Bitcoin decided to move to libsecp256k1 presumably because they like that it wasn't affected and that it has shown good testing and performance.

bitcoinj has some now old interface to libsecp256k1 but that diverged since; given Java bindings are now tested and available it may be a good idea to resume optionally supporting it in bitcoinj. @schildbach I've been toying with it and if you are interested could work on a PR.

> …presumably by the same key. This means massive address re-use, which bitcoinj already disregards by using HD wallets etc. So in a way, we implemented the countermeasure already years ago.

@schildbach True however with opt-in rbf and lightning we are likely to make multiple signatures with the same private key and the methods used to extract the key may be improved to the point where even low reuse is enough to extract a private key. If the private key is part of a bip32 wallet and public derived (non hardened, non private) and the attack has access to your public master then it is game over the attacker can reconstruct your private master and all your private keys, including privately derived are compromised.

The public master may be retrieved with a localised attack to the user, or using some sharing scheme (multisig) or by attacking a bip32 block explorer or one of these identify to address schemes/services.

Thoughts?

**Comment by schildbach** [comment]
Yes, I'd be happy with the libsecp256k1 JNI bindings being refreshed. A PR is much appreciated, thanks for offering to work on that!
